controllers/medusa/medusatask_controller.go

```diff
@@ -0,0 +1,470 @@
+					}
+					task.Status.Finished = append(task.Status.Finished, purgeResponse)
+				} else {
+					task.Status.Failed = append(task.Status.Failed, pod.Name)
+				}
+			}()
+		}
+		wg.Wait()
+		logger.Info("finished task operations")
+		if err := r.Status().Patch(context.Background(), task, patch); err != nil {
+			logger.Error(err, "failed to patch status", "MedusaTask", fmt.Sprintf("%s/%s", task.Spec.Operation, task.Namespace))
+		}
+	}()
+
+	return ctrl.Result{RequeueAfter: r.DefaultDelay}, nil
+}
+
+func (r *MedusaTaskReconciler) prepareRestoreOperation(ctx context.Context, task *medusav1alpha1.MedusaTask, pods []corev1.Pod, logger logr.Logger) (reconcile.Result, error) {
+	logger.Info("Starting prepare restore operations")
+
+	// Set the start time and add the pods to the in progress list
+	patch := client.MergeFrom(task.DeepCopy())
+	task.Status.StartTime = metav1.Now()
+	for _, pod := range pods {
+		task.Status.InProgress = append(task.Status.InProgress, pod.Name)
+	}
+
+	if err := r.Status().Patch(ctx, task, patch); err != nil {
+		logger.Error(err, "Failed to patch status")
+		// We received a stale object, requeue for next processing
+		return ctrl.Result{RequeueAfter: r.DefaultDelay}, err
+	}
+
+	// Invoke the purge operation on all pods, in the background
+	go func() {
+		wg := sync.WaitGroup{}
+
+		// Mutex to prevent concurrent updates to the backup.Status object
+		backupMutex := sync.Mutex{}
+		patch := client.MergeFrom(task.DeepCopy())
+
+		for _, p := range pods {
+			pod := p
+			wg.Add(1)
+			go func() {
+				logger.Info("starting prepare restore", "CassandraPod", pod.Name)
+				succeeded := false
+				err := prepareRestore(ctx, task, &pod, r.ClientFactory)
+				if err == nil {
+					logger.Info("finished prepare restore", "CassandraPod", pod.Name)
+					succeeded = true
+				} else {
+					logger.Error(err, "prepare restore failed", "CassandraPod", pod.Name)
+				}
+				backupMutex.Lock()
+				defer backupMutex.Unlock()
+				defer wg.Done()
+				task.Status.InProgress = utils.RemoveValue(task.Status.InProgress, pod.Name)
+				if succeeded {
+					operationResponse := medusav1alpha1.TaskResult{
+						PodName: pod.Name,
+					}
+					task.Status.Finished = append(task.Status.Finished, operationResponse)
+				} else {
+					task.Status.Failed = append(task.Status.Failed, pod.Name)
+				}
+			}()
+		}
+		wg.Wait()
+		logger.Info("finished task operations")
+		if err := r.Status().Patch(context.Background(), task, patch); err != nil {
+			logger.Error(err, "failed to patch status", "MedusaTask", fmt.Sprintf("%s/%s", task.Spec.Operation, task.Namespace))
+		}
+	}()
+
+	return ctrl.Result{RequeueAfter: r.DefaultDelay}, nil
+}
+
+func (r *MedusaTaskReconciler) syncOperation(ctx context.Context, task *medusav1alpha1.MedusaTask, pods []corev1.Pod, logger logr.Logger) (reconcile.Result, error) {
+	logger.Info("Starting sync operation")
+	syncStartTime := metav1.Now()
+	for _, pod := range pods {
```

Comment:
Won't we be doing duplicate work here by iterating over pods? It seems like we should first query for remote backups, which may involve queries across multiple pods, and then iterate across the backups. That's the pattern you have implemented for querying k8s backups in the following for loop.

Reply:
We're just iterating until we find one pod that answers back correctly (should be the first one).
We return as soon as the task is completed without moving to the next pod.
`GetBackups()` will return the backups from all nodes which is why we need a single pod to respond here.